cpp/src/parquet/encoding-benchmark.cc

```diff
@@ -163,4 +172,132 @@ static void BM_DictDecodingInt64_literals(benchmark::State& state) {
 
 BENCHMARK(BM_DictDecodingInt64_literals)->Range(1024, 65536);
 
+std::shared_ptr<::arrow::Array> MakeRandomStringsWithRepeats(size_t num_unique,
+                                                             size_t num_values) {
+  const int64_t min_length = 2;
+  const int64_t max_length = 10;
+  std::vector<uint8_t> buffer(max_length);
+  std::vector<std::string> dictionary(num_unique);
+
+  uint32_t seed = 0;
+  std::default_random_engine gen(seed);
+  std::uniform_int_distribution<int64_t> length_dist(min_length, max_length);
+
+  std::generate(dictionary.begin(), dictionary.end(), [&] {
+    auto length = length_dist(gen);
+    ::arrow::random_ascii(length, seed++, buffer.data());
+    return std::string(buffer.begin(), buffer.begin() + length);
+  });
+
+  std::uniform_int_distribution<int64_t> indices_dist(0, num_unique - 1);
+  ::arrow::StringBuilder builder;
+
+  for (size_t i = 0; i < num_values; i++) {
+    const auto index = indices_dist(gen);
+    const auto value = dictionary[index];
+    ABORT_NOT_OK(builder.Append(value));
+  }
+
+  std::shared_ptr<::arrow::Array> result;
+  ABORT_NOT_OK(builder.Finish(&result));
+  return result;
+}
+
+class BM_PlainDecodingByteArray : public ::benchmark::Fixture {
+ public:
+  void SetUp(const ::benchmark::State& state) override {
+    num_values_ = static_cast<int>(state.range());
+
+    input_array_ = MakeRandomStringsWithRepeats(num_values_ / 8, num_values_);
+    const auto& binary_array = static_cast<const ::arrow::BinaryArray&>(*input_array_);
+    values_ = std::vector<ByteArray>();
+    values_.reserve(num_values_);
+    total_size_ = 0;
+
+    for (int64_t i = 0; i < binary_array.length(); i++) {
+      auto view = binary_array.GetView(i);
+      values_.emplace_back(static_cast<uint32_t>(view.length()),
+                           reinterpret_cast<const uint8_t*>(view.data()));
+      total_size_ += view.length();
+    }
+
+    valid_bits_ =
+        std::vector<uint8_t>(::arrow::BitUtil::BytesForBits(num_values_) + 1, 255);
+
+    auto encoder = MakeTypedEncoder<ByteArrayType>(Encoding::PLAIN);
+    encoder->Put(values_.data(), num_values_);
+    buffer_ = encoder->FlushValues();
+  }
+
+  void TearDown(const ::benchmark::State& state) override {}
+
+ protected:
+  int num_values_;
+  std::shared_ptr<::arrow::Array> input_array_;
+  uint64_t total_size_;
+  std::vector<ByteArray> values_;
+  std::vector<uint8_t> valid_bits_;
+  std::shared_ptr<Buffer> buffer_;
+};
+
+BENCHMARK_DEFINE_F(BM_PlainDecodingByteArray, DecodeArrow_Dense)
+(benchmark::State& state) {
+  while (state.KeepRunning()) {
```

Comment:
From the [google/benchmark `README.md`](https://github.com/google/benchmark#a-faster-keeprunning-loop): 

> "In C++11 mode, a ranged-based for loop should be used in preference to the KeepRunning loop for running the benchmarks."

Apparently, the generated assembly code for the range-based for loop in this context is slightly more efficient. However, looking at [other existing Parquet benchmarks](https://github.com/apache/arrow/blob/75c83578ced6de80015f3b93c1d290b9e74430e5/cpp/src/parquet/arrow/reader-writer-benchmark.cc#L210), the `while (state.KeepRunning())` paradigm seems to be widely in use in the Arrow project, so I will leave it up to other Arrow community members to comment further on whether the range-based for loop convention is worth following instead.

Personally, I think `while (state.KeepRunning())` is slightly more readable.

Reply:
We're starting to use the range-based loop in new benchmarks. I recommend using that

Reply:
Thanks @wesm, I've updated the encoding benchmarks in the latest commit.